README.md
[![Yii2](https://img.shields.io/badge/required-Yii2_v2.0.40-blue.svg)](https://packagist.org/packages/yiisoft/yii2)
[![Downloads](https://img.shields.io/packagist/dt/wdmg/yii2-media.svg)](https://packagist.org/packages/wdmg/yii2-media)
[![Packagist Version](https://img.shields.io/packagist/v/wdmg/yii2-media.svg)](https://packagist.org/packages/wdmg/yii2-media)
![Progress](https://img.shields.io/badge/progress-ready_to_use-green.svg)
[![GitHub license](https://img.shields.io/github/license/wdmg/yii2-media.svg)](https://github.com/wdmg/yii2-media/blob/master/LICENSE)

<img src="./docs/images/yii2-media.png" width="100%" alt="Yii2 Media" />

# Yii2 Media
Media library for Yii2 with media categories and drag'n'drop uploader.

This module is an integral part of the [Butterfly.СMS](https://butterflycms.com/) content management system, but can also be used as an standalone extension.

Copyrights (c) 2019-2023 [W.D.M.Group, Ukraine](https://wdmg.com.ua/)

# Requirements 
* PHP 5.6 or higher
* Yii2 v.2.0.40 and newest
* [Yii2 Base](https://github.com/wdmg/yii2-base) module (required)
* [Yii2 Imagine](https://github.com/yiisoft/yii2-imagine) extension (required)
* [Yii2 SelectInput](https://github.com/wdmg/yii2-selectinput) widget

# Installation
To install the module, run the following command in the console:

`$ composer require "wdmg/yii2-media"`

After configure db connection, run the following command in the console:

`$ php yii media/init`

And select the operation you want to perform:
  1) Apply all module migrations
  2) Revert all module migrations

# Migrations
In any case, you can execute the migration and create the initial data, run the following command in the console:

`$ php yii migrate --migrationPath=@vendor/wdmg/yii2-media/migrations`

# Configure
To add a module to the project, add the following data in your configuration file:

    'modules' => [
        ...
        'media' => [
            'class' => 'wdmg\media\Module',
            'routePrefix' => 'admin',
            'mediaRoute' => '/media', // Routes to render media item (use "/" - for root)
            'mediaCategoriesRoute' => '/media/categories', // Routes to render media categories (use "/" - for root)
            'mediaPath' => '/uploads/media', // Path to save media files in @webroot
            'mediaThumbsPath' => '/uploads/media/_thumbs', // Path to save media thumbnails in @webroot
            'maxFilesToUpload' => 10, // maximum files to upload
            'maxUploadSize' => 5242880, // max file size in bytes to upload
            'allowedMime' => [ // allowed mime types
                'image/png' => true,
                'image/jpeg' => true,
                'image/gif' => true,
                'image/svg+xml' => true,
                'application/pdf' => true,
                'application/msword' => true,
                'application/vnd.ms-excel' => true,
                'application/rtf' => true,
                'text/csv' => true,
                'text/plain' => true,
                ...
            ]
        ],
        ...
    ],


# Routing
Use the `Module::dashboardNavItems()` method of the module to generate a navigation items list for NavBar, like this:

    <?php
        echo Nav::widget([
        'options' => ['class' => 'navbar-nav navbar-right'],
            'label' => 'Modules',
            'items' => [
                Yii::$app->getModule('media')->dashboardNavItems(),
                ...
            ]
        ]);
    ?>

# Status and version [ready to use]
* v.2.0.1 - Update dependencies
* v.2.0.0 - Some fix, update copyrights
* v.1.0.6 - RBAC implementation, read EXIF
* v.1.0.5 - Update README.md


CHANGELOG.md
Changelog
=========

## 2.0.1 (2023-07-07)
 * Update dependencies

## 2.0.0 (2023-06-25)
 * Some fix, update copyrights

## 1.0.6 (2021-01-02)
 * RBAC implementation
 * Read EXIF data
 * Maximum upload file size
 
## 1.0.5 (2020-06-14)
 * Update README.md
 
## 1.0.4 (2020-04-28)
 * Added drag-n-drop for upload
 * Update README.md and dependencies
 * Updating year in copyrights
 * Change icon class (FontAwesome update)
 
## 1.0.3 (2020-03-31)
 * Refactoring log of activity
 * Log activity

## 1.0.2 (2020-03-27)
 * Added pagination
 * Up to date dependencies
 
## 1.0.1 (2020-03-21)
 * Check file not exists or generate unique filename
 * Update README.md

## 1.0.0 (2020-03-20)
 * Added bulk actions for media items
 * CRUD for media items/categories, translations
 * Added file types icons
 
## 0.0.3 (2020-03-19)
 * Source and thumbnail fix
 * Image preview thumbnails
 * Added mime types validator, edit views for media items
 * Added upload functionality
 
## 0.0.2 (2020-03-15)
 * Migrations, base module and models
 
## 0.0.1 (2019-12-20)
 * Initial commit

Module.php
<?php

namespace wdmg\media;

/**
 * Yii2 Media library
 *
 * @category        Module
 * @version         2.0.1
 * @author          Alexsander Vyshnyvetskyy <alex.vyshnyvetskyy@gmail.com>
 * @link            https://github.com/wdmg/yii2-media
 * @copyright       Copyright (c) 2019 - 2023 W.D.M.Group, Ukraine
 * @license         https://opensource.org/licenses/MIT Massachusetts Institute of Technology (MIT) License
 *
 */

use wdmg\helpers\ArrayHelper;
use wdmg\helpers\StringHelper;
use Yii;
use wdmg\base\BaseModule;

/**
 * Media module definition class
 */
class Module extends BaseModule
{
    /**
     * {@inheritdoc}
     */
    public $controllerNamespace = 'wdmg\media\controllers';

    /**
     * {@inheritdoc}
     */
    public $defaultRoute = "list/index";

    /**
     * @var string, the name of module
     */
    public $name = "Media";

    /**
     * @var string, the description of module
     */
    public $description = "Media library";

    /**
     * @var string the module version
     */
    private $version = "2.0.1";

    /**
     * @var string the default routes to render media item (use "/" - for root)
     */
    public $mediaRoute = "/media";

    /**
     * @var string the default routes to render media categories (use "/" - for root)
     */
    public $mediaCategoriesRoute = "/media/categories";

    /**
     * @var string, the default path to save media files in @webroot
     */
    public $mediaPath = "/uploads/media";

    /**
     * @var string, the default path to save media thumbnails in @webroot
     */
    public $mediaThumbsPath = "/uploads/media/_thumbs";

    /**
     * @var int, maximum files to upload
     */
    public $maxFilesToUpload = 10;

    /**
     * @var int, max file size in bytes to upload
     */
    public $maxUploadSize = 5242880;

    /**
     * @var array, allowed mime types
     */
    public $allowedMime = [
        'image/png' => true,
        'image/jpeg' => true,
        'image/gif' => true,
        'image/tiff' => false,
        'image/svg+xml' => true,
        'image/webp' => false,
        'image/bmp' => false,
        'video/mpeg' => true,
        'video/mp4' => true,
        'video/ogg' => false,
        'video/quicktime' => true,
        'video/webm' => true,
        'video/x-matroska' => false,
        'video/x-ms-wmv' => true,
        'video/vnd.avi' => true,
        'audio/mpeg' => true,
        'audio/webm' => false,
        'audio/ogg' => false,
        'audio/mp4' => false,
        'audio/aac' => false,
        'audio/x-matroska' => false,
        'audio/x-ms-wma' => false,
        'audio/vnd.wave' => false,
        'audio/vnd.rn-realaudio' => false,
        'application/pdf' => true,
        'application/msword' => true,
        'application/vnd.oasis.opendocument.text' => false,
        'application/vnd.ms-excel' => true,
        'application/vnd.oasis.opendocument.spreadsheet' => false,
        'application/vnd.ms-powerpoint' => false,
        'application/vnd.oasis.opendocument.presentation' => false,
        'application/rtf' => true,
        'application/json' => true,
        'text/csv' => true,
        'text/plain' => true,
    ];

    /**
     * @var array, mime types collection of support
     */
    private $mimeTypes = [
        'image/png' => [
            'icon' => 'icon-image-png',
            'title' => 'PNG',
            'description' => 'Portable Network Graphics',
            'extensions' => ['.png', '.apng'],
            'type' => 'image'
        ],
        'image/jpeg' => [
            'icon' => 'icon-image-jpeg',
            'title' => 'JPEG',
            'description' => 'Joint Photographic Experts Group Format',
            'extensions' => ['.jpg', '.jpeg', '.jpe', '.jfif'],
            'type' => 'image'
        ],
        'image/gif' => [
            'icon' => 'icon-image-gif',
            'title' => 'GIF',
            'description' => 'Graphics Interchange Format',
            'extensions' => ['.gif'],
            'type' => 'image'
        ],
        'image/tiff' => [
            'icon' => 'icon-image-tiff',
            'title' => 'TIFF',
            'description' => 'Tagged Image File Format',
            'extensions' => ['.tif', '.tiff'],
            'type' => 'image'
        ],
        'image/svg+xml' => [
            'icon' => 'icon-image-svg',
            'title' => 'SVG',
            'description' => 'Scalable Vector Graphics',
            'extensions' => ['.svg', '.svgz'],
            'type' => 'image'
        ],
        'image/webp' => [
            'icon' => 'icon-image-webp',
            'title' => 'WebP',
            'description' => 'Web Picture Format',
            'extensions' => ['.webp'],
            'type' => 'image'
        ],
        'image/bmp' => [
            'icon' => 'icon-image-bmp',
            'title' => 'BMP',
            'description' => 'Windows OS/2 Bitmap Graphics',
            'extensions' => ['.bmp'],
            'type' => 'image'
        ],
        'video/mpeg' => [
            'icon' => 'icon-video-mpg',
            'title' => 'MPEG',
            'description' => 'Moving Picture Experts Group',
            'extensions' => ['.mpg', '.mpeg', '.mpv'],
            'type' => 'video'
        ],
        'video/mp4' => [
            'icon' => 'icon-video-mp4',
            'title' => 'MP4',
            'description' => 'MPEG-4 Part 14 Video Format',
            'extensions' => ['.mp4'],
            'type' => 'video'
        ],
        'video/ogg' => [
            'icon' => 'icon-video-ogv',
            'title' => 'OGG',
            'description' => 'Ogg Theora',
            'extensions' => ['.ogv', '.ogg'],
            'type' => 'video'
        ],
        'video/quicktime' => [
            'icon' => 'icon-video-mov',
            'title' => 'QuickTime',
            'description' => 'QuickTime Media Format',
            'extensions' => ['.mov'],
            'type' => 'video'
        ],
        'video/webm' => [
            'icon' => 'icon-video-webm',
            'title' => 'WebM',
            'description' => 'Web Media Video Format',
            'extensions' => ['.webm'],
            'type' => 'video'
        ],
        'video/x-ms-wmv' => [
            'icon' => 'icon-video-wmv',
            'title' => 'WMV',
            'description' => 'Windows Media Video',
            'extensions' => ['.wmv'],
            'type' => 'video'
        ],
        'video/vnd.avi' => [
            'icon' => 'icon-video-avi',
            'title' => 'AVI',
            'description' => 'Audio Video Interleave',
            'extensions' => ['.avi'],
            'type' => 'video'
        ],
        'video/x-matroska' => [
            'icon' => 'icon-video-mkv',
            'title' => 'MKV',
            'description' => 'Matroska Video File',
            'extensions' => ['.mkv'],
            'type' => 'video'
        ],
        'audio/mpeg' => [
            'icon' => 'icon-audio-mp3',
            'title' => 'MP3',
            'description' => 'MPEG Layer 3',
            'extensions' => ['.mp3'],
            'type' => 'audio'
        ],
        'audio/webm' => [
            'icon' => 'icon-audio-weba',
            'title' => 'WebM Audio',
            'description' => 'Web Media Audio Format',
            'extensions' => ['.weba'],
            'type' => 'audio'
        ],
        'audio/ogg' => [
            'icon' => 'icon-audio-ogg',
            'title' => 'Ogg Audio',
            'description' => 'Ogg Vorbis Audio Format',
            'extensions' => ['.ogg', '.oga', '.sb0'],
            'type' => 'audio'
        ],
        'audio/mp4' => [
            'icon' => 'icon-audio-mp4',
            'title' => 'MP4',
            'description' => 'MPEG-4 Part 14 Audio Format',
            'extensions' => ['.mp4'],
            'type' => 'audio'
        ],
        'audio/aac' => [
            'icon' => 'icon-audio-aac',
            'title' => 'AAC',
            'description' => 'Advanced Audio Coding',
            'extensions' => ['.aac'],
            'type' => 'audio'
        ],
        'audio/x-ms-wma' => [
            'icon' => 'icon-audio-wma',
            'title' => 'WMA',
            'description' => 'Windows Media Audio',
            'extensions' => ['.wma'],
            'type' => 'audio'
        ],
        'audio/vnd.wave' => [
            'icon' => 'icon-audio-wav',
            'title' => 'WAV',
            'description' => 'Waveform Audio File Format',
            'extensions' => ['.wav'],
            'type' => 'audio'
        ],
        'audio/vnd.rn-realaudio' => [
            'icon' => 'icon-audio-rm',
            'title' => 'RM',
            'description' => 'RealMedia',
            'extensions' => ['.rm', '.ram', '.rmvb'],
            'type' => 'audio'
        ],
        'audio/x-matroska' => [
            'icon' => 'icon-audio-mka',
            'title' => 'MKA',
            'description' => 'Matroska Audio File',
            'extensions' => ['.mka'],
            'type' => 'audio'
        ],
        'application/pdf' => [
            'icon' => 'icon-document-pdf',
            'title' => 'PDF',
            'description' => 'Adobe Portable Document Format',
            'extensions' => ['.pdf'],
            'type' => 'document'
        ],
        'application/msword' => [
            'icon' => 'icon-document-doc',
            'title' => 'MS Word',
            'description' => 'Microsoft Office Word',
            'extensions' => ['.doc', '.docx'],
            'type' => 'document'
        ],
        'application/vnd.oasis.opendocument.text' => [
            'icon' => 'icon-document-odt',
            'title' => 'ODF',
            'description' => 'OpenDocument Text',
            'extensions' => ['.odt'],
            'type' => 'document'
        ],
        'application/vnd.ms-excel' => [
            'icon' => 'icon-document-xls',
            'title' => 'MS Excel',
            'description' => 'Microsoft Office Excel',
            'extensions' => ['.xls', '.xlsx'],
            'type' => 'document'
        ],
        'application/vnd.oasis.opendocument.spreadsheet' => [
            'icon' => 'icon-document-ods',
            'title' => 'ODS',
            'description' => 'OpenDocument Spreadsheet',
            'extensions' => ['.ods'],
            'type' => 'document'
        ],
        'application/vnd.ms-powerpoint' => [
            'icon' => 'icon-document-ppt',
            'title' => 'MS PowerPoint',
            'description' => 'Microsoft Office PowerPoint',
            'extensions' => ['.ppt'],
            'type' => 'document'
        ],
        'application/vnd.oasis.opendocument.presentation' => [
            'icon' => 'icon-document-odp',
            'title' => 'ODP',
            'description' => 'OpenDocument Presentation',
            'extensions' => ['.odp'],
            'type' => 'document'
        ],
        'application/rtf' => [
            'icon' => 'icon-document-rtf',
            'title' => 'RTF',
            'description' => 'Rich Text Format',
            'extensions' => ['.rtf'],
            'type' => 'document'
        ],
        'application/json' => [
            'icon' => 'icon-document-json',
            'title' => 'JSON',
            'description' => 'JavaScript Object Notation',
            'extensions' => ['.json'],
            'type' => 'document'
        ],
        'text/csv' => [
            'icon' => 'icon-document-csv',
            'title' => 'CSV',
            'description' => 'Comma-Separated Values',
            'extensions' => ['.csv'],
            'type' => 'document'
        ],
        'text/plain' => [
            'icon' => 'icon-document-txt',
            'title' => 'TXT',
            'description' => 'Plain Text',
            'extensions' => ['.txt'],
            'type' => 'document'
        ]
    ];

    /**
     * @var integer, priority of initialization
     */
    private $priority = 5;

    /**
     * {@inheritdoc}
     */
    public function init()
    {
        parent::init();

        // Set version of current module
        $this->setVersion($this->version);

        // Set priority of current module
        $this->setPriority($this->priority);


        // Process and normalize route for media in frontend
        $this->mediaRoute = self::normalizeRoute($this->mediaRoute);
        $this->mediaCategoriesRoute = self::normalizeRoute($this->mediaCategoriesRoute);

        // Normalize path for media folder
        $this->mediaPath = \yii\helpers\FileHelper::normalizePath($this->mediaPath);
    }

    /**
     * {@inheritdoc}
     */
    public function dashboardNavItems($options = null)
    {
        $items = [
            'label' => $this->name,
            'url' => [$this->routePrefix . '/'. $this->id],
            'icon' => 'fa fa-fw fa-icons',
            'active' => in_array(\Yii::$app->controller->module->id, [$this->id]),
            'items' => [
                [
                    'label' => Yii::t('app/modules/media', 'Library'),
                    'url' => [$this->routePrefix . '/media/list/'],
                    'active' => (in_array(\Yii::$app->controller->module->id, ['content']) &&  Yii::$app->controller->id == 'list'),
                ],
                [
                    'label' => Yii::t('app/modules/media', 'Categories'),
                    'url' => [$this->routePrefix . '/media/cats/'],
                    'active' => (in_array(\Yii::$app->controller->module->id, ['content']) &&  Yii::$app->controller->id == 'cats'),
                ]
            ]
        ];

	    if (!is_null($options)) {

		    if (isset($options['count'])) {
			    $items['label'] .= '<span class="badge badge-default float-right">' . $options['count'] . '</span>';
			    unset($options['count']);
		    }

		    if (is_array($options))
			    $items = ArrayHelper::merge($items, $options);

	    }

	    return $items;
    }

    public function getAllowedMime($asArray = true)
    {
        $list = [];
        foreach ($this->allowedMime as $mime => $value) {
            if ($value === true)
                $list[] = $mime;
        }

        if ($asArray)
            return $list;
        else
            return implode(", ", $list);

    }

    public function getMimeTypes($asArray = true)
    {
        $list = $this->mimeTypes;

        if ($asArray)
            return $list;
        else
            return implode(", ", $list);

    }

    public function getAllowedExtensions($asArray = true)
    {
        $list = [];
        foreach ($this->allowedMime as $mime => $value) {
            if ($value === true && isset($types[$mime])) {
                if (isset($types[$mime]['extensions'])) {
                    if ($extensions = $types[$mime]['extensions']) {
                        foreach ($extensions as $extension) {
                            $list[] = $extension;
                        }
                    }
                }

            }
        }

        if ($asArray)
            return $list;
        else
            return str_replace(".", "", implode(", ", $list));

    }

    /**
     * @return array
     */
    public function getMimeTypesList($allTypes = false)
    {
        $list = [];
        if ($allTypes)
            $list['*'] = Yii::t('app/modules/media', 'All types');

        if ($types = $this->getMimeTypes()) {
            $allowed = $this->getAllowedMime();
            foreach ($types as $mime => $value) {
                if (in_array($mime, $allowed)) {
                    $list[$mime] = $value['title'];
                }
            }
        }

        return $list;
    }

    public function getTypeByMime($mime = null) {
        if (!is_null($mime)) {
            $types = $this->getMimeTypes();
            if (isset($types[$mime])) {
                if ($mime = $types[$mime]) {
                    return $mime;
                }
            }
        }

        return null;
    }

    /**
     * Returns the allowed maximum size of uploaded files.
     *
     * @param bool $formatted
     * @return int, size in bytes
     */
    public function getMaxUploadLimit($formatted = false)
    {
        $max_file_size = null;
        if (!$this->isConsole() && Yii::$app->request->isPost)
            $max_file_size = (int)Yii::$app->request->post('MAX_FILE_SIZE');

        $max_upload_limit = StringHelper::sizeToBytes(\ini_get('upload_max_filesize'));
        $max_post_limit = StringHelper::sizeToBytes(\ini_get('post_max_size'));

        if ($max_upload_limit > $max_post_limit)
            Yii::warning('PHP.ini\'s \'post_max_size\' is less than \'upload_max_filesize\'.', __METHOD__);

        if (isset($max_post_limit) && $max_post_limit < $max_upload_limit && $max_post_limit > 0)
            $limit = $max_post_limit;

        if (isset($this->maxUploadSize) && $this->maxUploadSize < $limit && $this->maxUploadSize > 0)
            $limit = $this->maxUploadSize;

        if (isset($max_file_size) && $max_file_size < $limit && $max_file_size > 0)
            $limit = $max_file_size;

        if ($formatted)
            return \Yii::$app->formatter->asShortSize($limit);

        return $limit;
    }

    /**
     * {@inheritdoc}
     */
    public function install()
    {
        parent::install();
        $path = Yii::getAlias('@webroot') . $this->mediaPath;
        if (\yii\helpers\FileHelper::createDirectory($path, $mode = 0775, $recursive = true))
            return true;
        else
            return false;
    }

    /**
     * {@inheritdoc}
     */
    public function uninstall()
    {
        parent::uninstall();
        $path = Yii::getAlias('@webroot') . $this->mediaPath;
        if (\yii\helpers\FileHelper::removeDirectory($path))
            return true;
        else
            return false;
    }
}